modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1038,350 +1071,107 @@ def _union_all(
         HdkOnNativeDataframe
             The new frame.
         """
-        frames = [self] + other_modin_frames
-
-        # This is a special case, where we need to preserve the index of empty frames.
-        if any(len(f.columns) == 0 for f in frames):
-            return self._union_all_arrow(frames, join, sort, ignore_index)
+        index_cols = None
+        columns = OrderedDict()
+        for col in self.columns:
+            columns[col] = self._dtypes[col]
 
-        # determine output columns
-        def join_cols():
-            new_cols_map = OrderedDict()
-            for col in self.columns:
-                new_cols_map[col] = self._dtypes[col]
+        if join == "inner":
             for frame in other_modin_frames:
-                if join == "inner":
-                    for col in list(new_cols_map):
-                        if col not in frame.columns:
-                            del new_cols_map[col]
-                elif join == "outer":
-                    for col in frame.columns:
-                        if col not in new_cols_map:
-                            new_cols_map[col] = frame._dtypes[col]
+                for col in list(columns):
+                    if col not in frame.columns:
+                        del columns[col]
+        elif join == "outer":
+            for frame in other_modin_frames:
+                for col in frame.columns:
+                    if col not in columns:
+                        columns[col] = frame._dtypes[col]
+        else:
+            raise NotImplementedError(f"Unsupported join type {join=}")
+
+        frames = []
+        for frame in [self] + other_modin_frames:
+            # Empty frames are filtered out only in case of the outer join.
+            if (
+                join == "inner"
+                or len(frame.columns) != 0
+                or (frame.has_index_cache and len(frame.index) != 0)
+                or (not frame.has_index_cache and frame.index_cols)
+            ):
+                if isinstance(frame._op, UnionNode):
+                    frames.extend(frame._op.input)
                 else:
-                    raise NotImplementedError(f"Unsupported join type {join=}")
-
+                    frames.append(frame)
+
+        if len(columns) == 0:
+            if len(frames) == 0:
+                dtypes = pd.Series()
+            elif ignore_index:
+                index_cols = [UNNAMED_IDX_COL_NAME]
+                dtypes = pd.Series([get_dtype(int)], index=index_cols)
+            else:
+                index_names = ColNameCodec.concat_index_names(frames)
+                index_cols = list(index_names)
+                dtypes = pd.Series(index_names.values(), index=index_cols)
+        else:
+            # Find common dtypes
```

Comment:
Why is this section called `Find common dtypes`, but only `columns` variable is changed?

Reply:
`columns` is a dictionary with the column names and dtypes.